storey/flow.py

```diff
@@ -386,6 +388,104 @@ async def _do(self, event):
                 await self._worker_awaitable
 
 
+class _PendingEvent:
+    def __init__(self):
+        self.in_flight = []
+        self.pending = []
+
+
+class _ConcurrentByKeyJobExecution(Flow):
+    def __init__(self, max_in_flight=8, **kwargs):
+        Flow.__init__(self, **kwargs)
+        self._max_in_flight = max_in_flight
+        self._q = None
+        self._pending_by_key = {}
+
+    async def _worker(self):
+        try:
+            while True:
+                job = await self._q.get()
+                if job is _termination_obj:
+                    for key, pending_event in self._pending_by_key.items():
+                        if self._pending_by_key[key].pending and not self._pending_by_key[key].in_flight:
+                            resp = await self._process_event(pending_event.pending[0])
+                            for event in pending_event.pending:
```

Comment:
Are we ok with this being O(N)?

Reply:
the `_handle_completed`? how can it not be O(n) if we want to send each of them downstream